static/flow/views/pi-selector-panel.js

```diff
@@ -741,6 +747,46 @@ var PiSelectorPanel = function(options) {
         clearTimeout(loadPiListTimer);
     }
 
+    //
+    // Start timer to determine if pi communication is taking too long
+    //
+    this.restartPiResponseTimer = function(runProgram, resetDeviceControls, refreshCallback) {
+        clearTimeout(piResponseTimer);
+        piResponseTimer = setTimeout(stopWaitingForResponse, piResponseTimerInterval, runProgram, resetDeviceControls, refreshCallback);
+    }
+
+    //
+    // Stop timer that determine if pi communication is taking too long
+    //
+    this.disablePiResponseTimer = function() {
+        clearTimeout(piResponseTimer);
+    }
+
+    //
+    // Timer fired, pi did not respond, stop listening and update state
+    //
+    function stopWaitingForResponse(runProgram, resetDeviceControls, refreshCallback) {
+        // Clear message handler
+        removeMessageHandlers();
+
+        modalAlert({title: 'Pi Timeout', message: 'Communication with Pi is taking longer than expected. Please try again. If issue continues, try rebooting Pi.', nextFunc: function() {
+            // Restore UI
+            if (runProgram) {
+                exitRunProgramState();
+                reselectCurrentPi();
+            } else {
+                if (typeof refreshCallback === "function") {
```

Comment:
It's your call but you might want to encapsulate this into a function with the `refreshCallback` as a parameter since and reuse it on your new line 708.